crates/driver/src/infra/bad_token/mod.rs

```diff
@@ -0,0 +1,354 @@
+                if let Some(params) = table.get("ThresholdBased") {
+                    Heuristic::ThresholdBased(params.threshold.into())
+                } else {
+                    Heuristic::default()
+                }
+            },
+            _ => Heuristic::default(),
+        };
+
+        let mode = match &config_file.mode {
+            Some(mode) if "LogOnly" == mode.as_str() => Mode::LogOnly,
+            Some(mode) if "Enabled" == mode.as_str() => Mode::Enabled,
+            _ => return Err(format!("unsupported mode")),
+        };
+
+        let config = Config {
+            solver,
+            supported_tokens,
+            unsupported_tokens,
+            timespan,
+            heuristic,
+            mode,
+        };
+
+        Ok(Self::from_config(config))
+    }
+
+    pub fn collect_from_path(file_path: &str, solvers: &Vec<Solver>) -> HashMap<Address, BadTokenMonitor> {
+        let paths = fs::read_dir(file_path).unwrap();
+
+        let mut bad_token_monitors = HashMap::new();
+        for path in paths {
+            if let Ok(file) = path {
+                let monitor = match Self::from_file_path(file.path()) {
+                    Ok(monitor) => monitor,
+                    Err(_) => continue,
+                };
+                
+                bad_token_monitors.insert(monitor.solver(), monitor);
+            }
+        }
+
+        let result = HashMap::new();
+
+        for solver in solvers {
+            match bad_token_monitors.remove(&solver.address()) {
+                Some(monitor) => result.insert(solver.address(), monitor),
+                None => result.insert(solver.address(), BadTokenMonitor::new(solver.address())),
+            }
+        }
+
+        result
+    }
+
+    pub async fn consolidate(&mut self, db: &Postgres, tokens: impl Iterator<Item = &TokenAddress>) -> Result<(), sqlx::Error> {
+        if Mode::LogOnly = self.config.mode() {
+            format!("implement logging");
+            return;
+        }
+
+        let mut tokens_to_insert = Vec::new();
+        let default_heuristic_state = HeuristicState::default(self.heuristic());
+        for token in tokens {
+            if self.explicit_allow_list.contains(token) {
+                continue;
+            }
+
+            if self.explicit_deny_list.contains(token) {
+                continue;
+            }
+
+            let mut heuristic_state = self
+                .heuristic_state
+                .get(token)
+                .map(|state| state.to_owned())
+                .unwrap_or(default_heuristic_state);
+
+            match (self.heuristic(), &mut heuristic_state) {
+                (
+                    Heuristic::ThresholdBased(threshold), 
+                    HeuristicState::ThresholdBased(threshold_state)
+                ) => {
+                    if threshold.threshold <= threshold_state.count + 1 {
+                        self.deny_list.insert(token.to_owned())
+                    }
```

Comment:
Unclear how the heuristic is supposed to work. Overall having doc comments (`///`) on all the relevant types would be helpful.
I'd imagine the heuristic to keep track of how often a token is involved in a solution that failed to simulate. Here it looks like it would just iterate over all the tokens in the auction without any indication whether they were good or not.